pkg/streamingpromql/engine_test.go

```diff
@@ -3292,100 +3306,641 @@ func TestQueryStats(t *testing.T) {
+				require.Equal(t, testCase.expectedTotalSamplesPerStep, prometheusSamplesStats.TotalSamplesPerStep, "invalid test case: expected per stepsamples does not match value from Prometheus' engine")
+			}
+
+			t.Run("without query planner", func(t *testing.T) {
+				mimirSamplesStats := runQueryAndGetSamplesStats(t, mimirEngine, testCase.expr, testCase.isInstantQuery)
+				require.Equal(t, testCase.expectedTotalSamples, mimirSamplesStats.TotalSamples)
+				require.Equal(t, testCase.expectedTotalSamplesPerStep, mimirSamplesStats.TotalSamplesPerStep)
+			})
+
+			t.Run("with query planner", func(t *testing.T) {
+				mimirSamplesStatsWithPlanning := runQueryAndGetSamplesStats(t, mimirEngineWithPlanning, testCase.expr, testCase.isInstantQuery)
+				require.Equal(t, testCase.expectedTotalSamples, mimirSamplesStatsWithPlanning.TotalSamples)
+				require.Equal(t, testCase.expectedTotalSamplesPerStep, mimirSamplesStatsWithPlanning.TotalSamplesPerStep)
+			})
+		})
+	}
+}
+
+func TestQueryStatsUpstreamTestCases(t *testing.T) {
+	// TestCases are taken from Prometheus' TestQueryStatistics.
+	opts := NewTestEngineOpts()
+	opts.CommonOpts.EnablePerStepStats = true
+	mimirEngine, err := NewEngine(opts, NewStaticQueryLimitsProvider(0), stats.NewQueryMetrics(nil), nil, log.NewNopLogger())
+	require.NoError(t, err)
+
+	prometheusEngine := promql.NewEngine(opts.CommonOpts)
+
+	planningOpts := opts
+	planningOpts.UseQueryPlanning = true
+	mimirEngineWithPlanning, err := NewEngine(planningOpts, NewStaticQueryLimitsProvider(0), stats.NewQueryMetrics(nil), NewQueryPlanner(planningOpts), log.NewNopLogger())
+	require.NoError(t, err)
+
+	storage := promqltest.LoadedStorage(t, `
+		load 10s
+		  metricWith1SampleEvery10Seconds 1+1x100
+		  metricWith3SampleEvery10Seconds{a="1",b="1"} 1+1x100
+		  metricWith3SampleEvery10Seconds{a="2",b="2"} 1+1x100
+		  metricWith3SampleEvery10Seconds{a="3",b="2"} 1+1x100
+		  metricWith1HistogramEvery10Seconds {{schema:1 count:5 sum:20 buckets:[1 2 1 1]}}+{{schema:1 count:10 sum:5 buckets:[1 2 3 4]}}x100
+	`)
+	t.Cleanup(func() { storage.Close() })
+
+	runQueryAndGetSamplesStats := func(t *testing.T, engine promql.QueryEngine, expr string, start, end time.Time, interval time.Duration) *promstats.QuerySamples {
+		var q promql.Query
+		var err error
+		opts := promql.NewPrometheusQueryOpts(true, 0)
+
+		if interval == 0 {
+			// Instant query
+			q, err = engine.NewInstantQuery(context.Background(), storage, opts, expr, start)
+		} else {
+			// Range query
+			q, err = engine.NewRangeQuery(context.Background(), storage, opts, expr, start, end, interval)
+		}
+
+		require.NoError(t, err)
+		defer q.Close()
+
+		res := q.Exec(context.Background())
+		require.NoError(t, res.Err)
+
+		return q.Stats().Samples
+	}
+
+	cases := []struct {
+		Query                       string
+		Start                       time.Time
+		End                         time.Time
+		Interval                    time.Duration
```

Comment:
[nit] We've tended not to expose these fields in other similar test cases:

```suggestion
		query                       string
		start                       time.Time
		end                         time.Time
		interval                    time.Duration
```